### PR TITLE
Add create test release script

### PR DESCRIPTION
**Pull request by diegocurbelo**
# Changes proposed in this Pull Request:

This PR add a bash script to the bin folder, to automate the creation of a test release version.

Before doing executing any steps a process summary and a confirmation to continue will be displayed

```
This script will perform the following actions:
  - Create a new branch named 'release/5.6.0' using 'develop' as base, and push it to GitHub
  - Run Woorelease simulation locally with version=5.6.0 (using the 'release/5.6.0' branch)
  - Create a new tag named '5.6.0-test' in GitHub
  - Create a new release named '5.6.0-test' in GitHub (marked as 'pre-release')
  - Upload the generated plugin zip file (woocommerce-gateway-stripe.zip) to the newly created release in GitHub

Do you want to continue? (yes/no)
```

Next steps:
- Convert it to a javascript task
- Read the GitHub token from the Woorelease config

# Testing instructions

To test it without changing this repo, fork it and change the GITHUB_ACCOUNT variable at the top of the script.

__NOTE__: the script will be removed during the process (when switching branches), that is expected 😄 

- Execute the script with version number  5.9.0
  - `./bin/create-test-release.sh 5.9.0`

- Follow the script instructions
  - To test the automatic install of Woorelease a GitHub token with repo access is required, if left empty instructions to manually install it will be displayed
  - To test the automatic release creation and asset upload a GitHub token is required, if left empty instructions to manually create the release will be displayed

- Verify that the branch `release/5.9.0` was created
  - In this branch verify that both `readme.txt` and `changelog.txt` have the correct version number (5.x.x replaced with 5.9.0)... the date should remain as 2021-xx-xx as that will be changed by woorelease at the time of the production release.
  
- Verify that the tag `5.9.0-test` was created

- Extract the generated file `woocommerce-gateway-stripe.zip` and verify the version numbers in:
  - woocommerce-gateway-stripe.php:8
  - readme.txt:7,129
  - changelog.txt:3
 
#### Test cleanup:
- Delete the branch `release/5.9.0`
- Delete the tag `5.9.0-test`

### CODE REVIEW

File: bin/create-test-release.sh [hunk 1]
```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env bash
+
```
Comment (reykjalin) [inline bin/create-test-release.sh:2]: Let's enable the so-called ["strict mode"](http://redsymbol.net/articles/unofficial-bash-strict-mode/)

```bash
set -euo pipefail
IFS=$'\n\t'
```

File: bin/create-test-release.sh [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	# We simulate creating the release with the version number VERSION (ex: 5.3.0) and not the current test version (ex: 5.3.0-test)
+	echo ''
+	echo "> php .cache/woorelease.phar simulate --product_version=$VERSION https://github.com/woocommerce/woocommerce-gateway-stripe/tree/release/$VERSION"
+	output=$(php .cache/woorelease.phar simulate --product_version=$VERSION https://github.com/woocommerce/woocommerce-gateway-stripe/tree/release/$VERSION | tee /dev/stderr )
+
+	# Update the 'Version' tag in woocommerce-gateway-stripe.php to the current version + '-test'
+	ZIP_FILE=$(echo $output | sed -n "s/.*Skipping upload of asset \(.*\) to GH release.*/\1/p")
+	[ $? -eq 0 ] || exit 1
+
+	cd $(echo "$ZIP_FILE" | rev | cut -d '/' -f3- | rev)
```
Comment (reykjalin) [inline bin/create-test-release.sh]: Let's just use `cd $(dirname "$ZIP_FILE")` 🙂

File: bin/create-test-release.sh [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	exit 0
+fi
+
+create_git_branch;
+create_test_package;
+
+echo ''
+echo "Test release $VERSION generated successfully."
+echo ''
+echo "You need to upload the zip package 'woocommerce-gateway-stripe.zip' to the tag '$VERSION-test' on GitHub (creating a version and marking it as pre-release BEFORE saving)."
```
Comment (reykjalin) [inline bin/create-test-release.sh]: I wonder if we could create the release and upload the zip file via the script here 🤔 

That'd be pretty sweet 👌

**Review by reykjalin — COMMENTED** [review]
I haven't tried running this yet, but it looks good! I see a couple things we can improve though 🙂 

I'll do some more testing for this tomorrow 👍

File: bin/create-test-release.sh (hunk 3 above)
Comment (diegocurbelo) [reply bin/create-test-release.sh]: Copied the release creation and asset upload form the WC Pay repo... and also added manual instructions as a fallback

File: bin/create-test-release.sh [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+	echo ''
+    echo "> git checkout -b release/$VERSION"
+	git checkout -b release/$VERSION
+
+    echo ''
+    echo "> sed -i '' \"s/^= 5.x.x - \(.*\)$/= $VERSION - \1/\" readme.txt"
+    sed -i '' "s/^= 5.x.x - \(.*\)$/= $VERSION - \1/" readme.txt
+    echo "> sed -i '' \"s/^= 5.x.x - \(.*\)$/= $VERSION - \1/\" changelog.txt"
+    sed -i '' "s/^= 5.x.x - \(.*\)$/= $VERSION - \1/" changelog.txt
```
Comment (reykjalin) [inline bin/create-test-release.sh]: These `sed` commands will not work when we move to newer major versions, so I think we need to change them to use `\d+` in place of `5` or something along those lines 🤔 

This is also very brittle. For example, `woorelease` fails if the `5.x.x` changelog entry is missing from either `changelog.txt` or `readme.txt`.

I'm not sure what the best way to fix that is 🤔 
I guess the release lead can make sure the changelog entries are correct in `develop` before running the script? That sort of defeats the point of it though...

Maybe we include a step before running `woorelease` where the script stops and instructs you to "make sure changelog entries are right" or something along those lines. Once you hit `<Enter>` the script continues?

We could also split this into 2 steps; 1 for preparing the branch automatically if possible, and another for creating the release once the branch has been prepared properly?

I'm not sure what the best way is to address this 🤔

File: bin/create-test-release.sh [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+check_prerequisites
+
+# Check if the 'release/VERSION' branch exists, and the next test SUFFIX to use (ex: 5.3.0-test3)
+check_github_status
+
+# Print the actions to perform
+print_summary
+
+echo ''
+read -p "Do you want to continue? (yes/no) " -n 1 -r
```
Comment (reykjalin) [inline bin/create-test-release.sh]: The `yes/no` should be `y/N`; as soon as you press <kbd>y</kbd> it continues, any other letter and it stops.

**Review by reykjalin — CHANGES_REQUESTED** [review]
Running the script doesn't really work, and I think it's because the approach is a bit brittle. No fault of yours, the changelogs and version numbers just have to be set up in a very specific way for `woorelease` to work 🤔 

Let's discuss that a bit in the PR 👍 

Great work so far!

File: bin/create-test-release.sh (hunk 5 above)
Comment (diegocurbelo) [reply bin/create-test-release.sh]: Changed

File: bin/create-test-release.sh (hunk 4 above)
Comment (diegocurbelo) [reply bin/create-test-release.sh]: Added logic to check for local and remote branches already created:
- if a remote branch exists... process starts with woorelease from that branch
- if a remote branch does NOT exist, but a local does... instructions are shown to check it, push it to github and re-run the script
-  if a neither a remote nor local branch exist... a new branch based on develop will be created, and an attempt to update the version number in both readme.txt and changelog.txt will be made, after that,  both files will be checked for the tags Woorelease expect (`= 5.4.0 - 2021-xx-xx =`), if they are not correct instructions are shown to check it, push it to github and re-run the script

```
Checking GitHub project status...

A local branch named release/5.6.0 already exists, but has not been pushed to the remote repository.

Make sure that both readme.txt and changelog.txt have the correct entries for the release version:
  = 5.6.0 - 2021-xx-xx =
Push the branch to the remote repository, and re-run this script.

Aborted.
```

Also, added an explicit `Aborted` message when exiting with an error, and the message `Partially completed, additional manual steps required` is shown when the release needs to be manually created in GitHub and the zip file uploaded:
```
You need to manually create a release in GitHub:
  - Go to: https://github.com/woocommerce/woocommerce-gateway-stripe/releases/new
  - Fill the form with the following values:
    Tag version   = 5.6.0-test
    Target        = release/5.6.0
    Release title = Version 5.6.0-test. Not for production.
    Description   = This version is for internal testing. It should NOT be used for production.
  - Attach the plugin zip file 'woocommerce-gateway-stripe.php'
  - Select the checkbox 'This is a pre-release'
  - Publish the release

Partially completed, additional manual steps required.
```

File: bin/create-test-release.sh [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+		echo "PHP is not installed"
+		abort
+	}
+
+	# Check for Woorelease phar file
+	if [[ ! -f ".cache/woorelease.phar" ]]; then
+		mkdir -p .cache
+
+		echo "Woorelease not found."
+		read -p "To automatically download it, paste your GitHub token: " -r
```
Comment (asumaran) [inline bin/create-test-release.sh:51]: The script is asking me for a GitHub token even though I've already added it to `GITHUB_TOKEN` lines above. I think we should check if `GITHUB_TOKEN` is already populated before asking for a token.

**Review by asumaran — COMMENTED** [review]
@diegocurbelo Sorry to chime in; I wanted to look at this as well. I've encountered a couple of issues while testing this PR.

1. After modifying the `create-test-release.sh` file with my `GITHUB_ACCOUNT` and `GITHUB_TOKEN` I wasn't able to run the script because of the following:

```
~/www/asumaran-stripe ‹tweak/add-script-to-create-test-release*› » ./bin/create-test-release.sh 5.9.0
Checking GitHub project status...
You have unstaged/uncommited changes, please commit or stash them, and then re-run this script.
Aborted.
```
2. After committing the changes (which I think it shouldn't be necessary) I was able to run the script but got another error: 

```
WR [21:00:11] [NOTICE] Attempting to bump main plugin version in readme.txt...
WR [21:00:11] [ERROR] Can't find the changelog version declaration in the readme.txt for woocommerce-gateway-stripe
WR [21:00:11] [ERROR] Woorelease finished with errors! Please correct the error and retry.
WR [21:00:11] [ERROR] For more information on the error, please visit https://github.com/woocommerce/woorelease/wiki/Errors#error-43 or ping #woo-devs in Slack
```

**Pull request closed without merging**